File: src/main/java/org/mitallast/queue/raft/persistent/ReplicatedLog.java

package org.mitallast.queue.raft.persistent;

import javaslang.collection.Vector;
import javaslang.control.Option;
import org.mitallast.queue.raft.protocol.LogEntry;
import org.mitallast.queue.raft.protocol.RaftSnapshot;

import java.io.Closeable;

public interface ReplicatedLog extends Closeable {

    boolean isEmpty();

    boolean contains(LogEntry entry);

    Vector<LogEntry> entries();

    int committedEntries();

    long committedIndex();

    boolean containsMatchingEntry(long otherPrevTerm, long otherPrevIndex);

    Option<Long> lastTerm();

    long lastIndex();

    long prevIndex();

    long nextIndex();

    ReplicatedLog commit(long committedIndex);

    ReplicatedLog append(LogEntry entry);

    ReplicatedLog append(Vector<LogEntry> entries);

    ReplicatedLog compactWith(RaftSnapshot snapshot);

    Vector<LogEntry> entriesBatchFrom(long fromIncluding, int howMany);

    Vector<LogEntry> slice(long from, long until);

    boolean containsEntryAt(long index);

    long termAt(long index);

    boolean hasSnapshot();

    RaftSnapshot snapshot();

    @Override
    void close();
}

File: src/main/java/org/mitallast/queue/raft/persistent/ReplicatedLog.kt
package org.mitallast.queue.raft.persistent

import javaslang.collection.Vector
import javaslang.control.Option
import org.mitallast.queue.raft.protocol.LogEntry
import org.mitallast.queue.raft.protocol.RaftSnapshot

import java.io.Closeable

interface ReplicatedLog : Closeable {

    val isEmpty: Boolean

    operator fun contains(entry: LogEntry): Boolean

    fun entries(): Vector<LogEntry>

    fun committedEntries(): Int

    fun committedIndex(): Long

    fun containsMatchingEntry(otherPrevTerm: Long, otherPrevIndex: Long): Boolean

    fun lastTerm(): Option<Long>

    fun lastIndex(): Long

    fun prevIndex(): Long

    fun nextIndex(): Long

    fun commit(committedIndex: Long): ReplicatedLog

    fun append(entry: LogEntry): ReplicatedLog

    fun append(entries: Vector<LogEntry>): ReplicatedLog

    fun compactWith(snapshot: RaftSnapshot): ReplicatedLog

    fun entriesBatchFrom(fromIncluding: Long, howMany: Int): Vector<LogEntry>

    fun slice(from: Long, until: Long): Vector<LogEntry>

    fun containsEntryAt(index: Long): Boolean

    fun termAt(index: Long): Long

    fun hasSnapshot(): Boolean

    fun snapshot(): RaftSnapshot

    override fun close()
}
